### PR TITLE
(PDB-2487) implement a psuedo resource_events_ttl

### PR DESCRIPTION
**Pull request by tkishel**

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -3,6 +3,7 @@
   Boolean $manage_database_maintenance = true,
   Boolean $manage_postgresql_settings  = true,
   Boolean $manage_table_settings       = true,
+  Integer $resource_events_ttl         = 0,
```
Comment (npwalker) [inline manifests/init.pp]: Either the resource_events_ttl class should be included here in init.pp and not in maintenance... OR this parameter shouldn't be here and should only be on maintenance

File: manifests/maintenance/resource_events_ttl.pp [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# A description of what this class does
+#
+# @summary A short summary of the purpose of this class
+#
+# @example
+#   include pe_databases::maintenance::manage_resource_events
+class pe_databases::maintenance::manage_resource_events (
```
Comment (npwalker) [inline manifests/maintenance/resource_events_ttl.pp]: class name doesn't match the file name

Comment (tkishel) [reply manifests/maintenance/resource_events_ttl.pp]: Gah!

File: manifests/init.pp (hunk 1 above)
Comment (tkishel) [reply manifests/init.pp]: Ok

**Comment by tkishel** [comment]
> Should include a spec of the new class to make sure it compiles

Done

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 - Make pg_repack the default way to perform maintenance [#25](https://github.com/npwalker/pe_databases/pull/25)
 - Start maintaining the reports table which we could not afford to perform a VACUUM FULL on
-- Rename typoed `pe_databases::maintenance::disable_maintenace` parameter to `pe_databases::maintenance::disable_maintenance`
+- Rename typoed `pe_databases::maintenance::disable_maintenance` parameter to `pe_databases::maintenance::disable_maintenance`
```
Comment (npwalker) [inline CHANGELOG.md]: ummm you fixed the typo but the typo needs to stay

Comment (tkishel) [reply CHANGELOG.md]: Barf

**Comment by tkishel** [comment]
> probably a spec to ensure setting the ttl to a positive integer results in the class being included

Seems a bit extreme, at that is actually testing Puppet class parameters and an if statement, but the next commit will do that.

**Comment by tkishel** [comment]
None of the other maintenance classes have similar tests, so modeled them ...

```
[root@pe-201911-master site-modules]# grep resource_events_ttl /etc/puppetlabs/code/environments/production/data/common.yaml


[root@pe-201911-master site-modules]# puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Info: Caching catalog for pe-201911-master.puppetdebug.vlan
Info: Applying configuration version 'pe-201911-master-production-ad49fff4b90'
Notice: Applied catalog in 33.40 seconds

[root@pe-201911-master site-modules]# crontab -l | grep resource_events
```

```
[root@pe-201911-master site-modules]# echo "pe_databases::maintenance::resource_events_ttl: 3" >> /etc/puppetlabs/code/environments/production/data/common.yaml 

[root@pe-201911-master site-modules]# grep resource_events_ttl /etc/puppetlabs/code/environments/production/data/common.yaml 
pe_databases::maintenance::resource_events_ttl: 3

[root@pe-201911-master site-modules]# puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Info: Caching catalog for pe-201911-master.puppetdebug.vlan
Info: Applying configuration version 'pe-201911-master-production-ad49fff4b90'
Notice: /Stage[main]/Pe_databases::Maintenance::Manage_resource_events/File[/opt/puppetlabs/pe_databases/scripts/delete_from_resource_events.sh]/ensure: defined content as '{md5}4f3b635d2a335037ca3d8ed8352b7768'
Notice: /Stage[main]/Pe_databases::Maintenance::Manage_resource_events/Cron[DELETE FROM resource_events]/ensure: created
Notice: Applied catalog in 21.37 seconds

[root@pe-201911-master site-modules]# crontab -l | grep resource_events
# Puppet Name: DELETE FROM resource_events
15 4 * * * /opt/puppetlabs/pe_databases/scripts/delete_from_resource_events.sh 3 > /var/log/puppetlabs/pe_databases_cron/output.log 2>&1
```

**Comment by tkishel** [comment]
Added spec tests, but they could be dry-er.

File: manifests/init.pp [hunk 4]
```diff
@@ -7,7 +7,9 @@
   String  $scripts_dir                 = "${install_dir}/scripts"
 ) {
 
-  if ( versioncmp('2017.3.0', $facts['pe_server_version']) <= 0 ) {
+  if ( versioncmp('2019.2.0', $facts['pe_server_version']) <= 0 ) {
+    $psql_version = '11'
```
[the remark this answers is not in the document]
Comment (npwalker) [reply manifests/init.pp]: yea the section of the if you added should be `$psql_version = $facts['pe_postgresql_info']['installed_server_version']`

[the remark this answers is not in the document]
Comment (npwalker) [reply manifests/init.pp]: Why would it break 2016.4?  The if you added only matches on 2019.2+

[the remark this answers is not in the document]
Comment (tkishel) [reply manifests/init.pp]: The `pe_postgresql_info` fact in general doesn't exist in 2016.4 tho I guess/hope `versioncmp` will return false and we end up in the else.

[the remark this answers is not in the document]
Comment (tkishel) [reply manifests/init.pp]: Or is this what you are looking for:

```
  if $facts['pe_postgresql_info'] {
    $psql_version = $facts['pe_postgresql_info']['installed_server_version']
  } else {
    $psql_version = '9.4'
  }
```

**Comment by tkishel** [comment]
Also, in PE 2016.4, the cron resources are not idempotent. Maybe a review of what is supported is due ...

```
[root@pe-2016415-master pe_databases]# puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for pe-2016415-master.puppetdebug.vlan
Info: Applying configuration version '1572455958'
Notice: /Stage[main]/Pe_databases::Backup/Cron[puppet_enterprise_database_backup_[pe-activity, pe-classifier, pe-postgres, pe-rbac, pe-orchestrator]]/minute: minute changed ['30'] to '30'
Notice: /Stage[main]/Pe_databases::Backup/Cron[puppet_enterprise_database_backup_[pe-activity, pe-classifier, pe-postgres, pe-rbac, pe-orchestrator]]/hour: hour changed ['22'] to '22'
Notice: /Stage[main]/Pe_databases::Backup/Cron[puppet_enterprise_database_backup_[pe-puppetdb]]/minute: minute changed ['0'] to '0'
Notice: /Stage[main]/Pe_databases::Backup/Cron[puppet_enterprise_database_backup_[pe-puppetdb]]/hour: hour changed ['2'] to '2'
Notice: /Stage[main]/Pe_databases::Backup/Cron[puppet_enterprise_database_backup_[pe-puppetdb]]/weekday: weekday changed ['7'] to '7'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL facts tables]/minute: minute changed ['30'] to '30'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL facts tables]/hour: hour changed ['4'] to '4'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL facts tables]/weekday: weekday changed ['2', '6'] to '2 6'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL catalogs tables]/minute: minute changed ['30'] to '30'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL catalogs tables]/hour: hour changed ['4'] to '4'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL catalogs tables]/weekday: weekday changed ['0', '4'] to '0 4'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL other tables]/minute: minute changed ['30'] to '30'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL other tables]/hour: hour changed ['5'] to '5'
Notice: /Stage[main]/Pe_databases::Maintenance::Vacuum_full/Cron[VACUUM FULL other tables]/monthday: monthday changed ['20'] to '20'
Notice: /Stage[main]/Pe_databases::Maintenance::Manage_resource_events/Cron[DELETE FROM resource_events]/minute: minute changed ['15'] to '15'
Notice: /Stage[main]/Pe_databases::Maintenance::Manage_resource_events/Cron[DELETE FROM resource_events]/hour: hour changed ['4'] to '4'
Notice: /Stage[main]/Puppet_enterprise::Mcollective::Server::Facter/Cron[pe-mcollective-metadata]/minute: minute changed ['0', '15', '30', '45'] to '0 15 30 45'
Notice: Applied catalog in 14.48 seconds
```

File: manifests/init.pp (hunk 4 above)
[the remark this answers is not in the document]
Comment (npwalker) [reply manifests/init.pp]:
```
if ( versioncmp('2019.2.0', $facts['pe_server_version']) <= 0 ) {
    $psql_version = $facts['pe_postgresql_info']['installed_server_version']
} elsif ( versioncmp('2017.3.0', $facts['pe_server_version']) <= 0 ) {
    $psql_version = '9.6'
} else {
    $psql_version = '9.4'
}
```

[the remark this answers is not in the document]
Comment (tkishel) [reply manifests/init.pp]: Yup

**Comment by npwalker** [comment]
psql_verison isn't an integer

**Comment by tkishel** [comment]
Sigh, I meant Float.

The value being assigned to `$psql_version` via `facts['pe_postgresql_info']['installed_server_version']` is being interpreted as an Float, which caused a type error in Travis:

```
 1) pe_databases::maintenance::pg_repack on debian-6-x86_64 on > PE 2019.1.0 is expected to contain Cron[pg_repack reports tables] with command => "su - pe-postgres -s /bin/bash -c \"/opt/puppetlabs/server/apps/postgresql/9.6/bin/pg_repack -d pe-puppetdb --jobs 2 -t reports -t resource_events\" > /var/log/puppetlabs/pe_databases_cron/output.log 2>&1"
     Failure/Error: include pe_databases::backup
     
     Puppet::PreformattedError:
       Evaluation Error: Error while evaluating a Function Call, Class[Pe_databases::Backup]: parameter 'psql_version' expects a String value, got Float (file: /home/travis/build/npwalker/pe_databases/spec/fixtures/modules/pe_databases/manifests/init.pp, line: 42, column: 5) on node travis-job-d0975df9-68d0-491e-bc3a-ef707652c56f.c.travis-ci-prod-2.internal
```
So we need to interpolate the fact to convert it to a String, and tell puppet-lint that doing so it okay.

**Comment by tkishel** [comment]
What do you think? y/n?

**Comment by tkishel** [comment]
Is there anything else you'd like, or can this be merged, or should it be closed?

**Comment by tkishel** [comment]
@npwalker Please advice as to whether to merge or close. Thx!

**Comment by tkishel** [comment]
It seems `resource_events_ttl` is a problem for PE HA.
Closing.

**Pull request closed without merging**